File: index/opus_sentid_index.py
#!/usr/bin/env python3


import argparse
import xml.parsers.expat
import zipfile
import os
import urllib.request
import sqlite3
import sys
from os.path import exists



storage_url_base = 'https://object.pouta.csc.fi/OPUS-'


parser = argparse.ArgumentParser(prog='opus_sentid_index',
    description='Read OPUS documents and extract sentence IDs and row numbers from an index DB')

parser.add_argument('-d', '--database', help='Database', required=True)
parser.add_argument('-c', '--corpus', help='Corpus name', required=True)
parser.add_argument('-r', '--release', help='Release version', required=True)
parser.add_argument('-l', '--language', help='Language', required=True)
parser.add_argument('-f', '--file_name',help='File name (if not given, prints all files)')
parser.add_argument('-v', '--verbose', help='verbose output', action='store_true', default=False)

args = parser.parse_args()

verbose = args.verbose;

corpus    = args.corpus
release   = args.release
data_url  = storage_url_base + corpus + '/' + release + '/raw/' + args.language + '.zip'
data_file = corpus + '_' + release + '_raw_' + args.language + '.zip'



if not exists(data_file):
    sys.stderr.write(f"Downloading {data_url}\n")
    urllib.request.urlretrieve(data_url, data_file)
    
lzip = zipfile.ZipFile(data_file)


def start_element(name, attrs):
    global inSent, sentStr, sentCount, sentID
    global parCount, parID
              
    if name == 's':
        inSent = True
        sentCount += 1
        sentStr = ''
        if 'id' in attrs:
            sentID = attrs['id']
        else:
            sentID = str(sentCount)
        if not sentCount % 2000:
            sys.stderr.write('.')
            if not sentCount % 100000:
                sys.stderr.write(f" {sentCount}\n")
            sys.stderr.flush()

    elif name == 'p' or name == 'P':
        parCount += 1
        if 'id' in attrs:
            parID = attrs['id']
        else:
            parID = str(sentCount)


            
def end_element(name):
    global inSent, sentStr, sentID
    global parStr, parID
    global corpus, release, document
    global cur, con, verbose
        
    if name == 's':
        sentStr = sentStr.lstrip().rstrip()
        res = cur.execute("""SELECT ROWID FROM sentences WHERE sentence = ?""", [sentStr])
        record = res.fetchone()
        if record:
            print("\t".join([str(record[0]),corpus,release,document,str(parID),str(sentID),str(len(sentStr))]))
        else:
            ## insert a new sentence!
            if verbose:
                sys.stderr.write('NEW SENTENCES - ' + sentID + ': ' + sentStr + "\n")
            cur.execute("""INSERT OR IGNORE INTO sentences VALUES(?)""", [sentStr])
            con.commit()                
            res = cur.execute("""SELECT ROWID FROM sentences WHERE sentence = ?""", [sentStr])
            record = res.fetchone()
            if record:
                print("\t".join([str(record[0]),corpus,release,document,str(parID),str(sentID),str(len(sentStr))]))
            else:
                sys.stderr.write('FAILED TO INSERT - ' + sentID + ': ' + sentStr + "\n")

    elif name == 'p' or name == 'P':
        parID = 0

        sentStr = ''
        inSent = False


def char_data(data):
    global inSent
    global sentStr
    if inSent:
        sentStr = sentStr + data


## wait for max 2 hours
# con = sqlite3.connect(args.database, timeout=7200, isolation_level='EXCLUSIVE')
con = sqlite3.connect(args.database, timeout=7200)
con.execute("PRAGMA journal_mode=WAL")
cur = con.cursor()

cur.execute("CREATE TABLE IF NOT EXISTS sentences ( sentence TEXT UNIQUE PRIMARY KEY NOT NULL )")

count = 0

for filename in lzip.namelist():
    if filename[-4:] == '.xml':

        parser = xml.parsers.expat.ParserCreate()
        parser.StartElementHandler = start_element
        parser.EndElementHandler = end_element
        parser.CharacterDataHandler = char_data

        inSent    = False
        sentStr   = ''
        sentID    = ''
        sentCount = 0
        parCount  = 0
        parID     = 0

        errorCount = 0
        
        document  = '/'.join(filename.split('/')[2:])

        if verbose:
            sys.stderr.write(f"process {filename} ({count} sentences done)\n")
        with lzip.open(filename, 'r') as f:
            for line in f:
                try:
                    parser.Parse(line)
                except:
                    errorCount += 1
                    if verbose:
                        sys.stderr.write(f"error parsing {line}\n")

        count += sentCount
        if errorCount > 0:
            sys.stderr.write(f"XML parsing errors for {filename}: {errorCount} lines\n")


sys.stderr.write(f"A total of {count} sentences found\n")
os.unlink(data_file)

File: index/opus_sentid_index.pl
#!/usr/bin/env perl

use strict;

use MCDB_File;
use XML::Parser;
use Archive::Zip::SimpleUnzip qw($SimpleUnzipError) ;

use vars qw($opt_c $opt_d $opt_l $opt_r);
use Getopt::Std;
getopts('c:d:l:r:');

my $database = $opt_d || die "need to specify a sent2id database file";
my $corpus = $opt_c   || die "need to specify a valid corpus name";
my $release = $opt_r  || die "need to specify a valid corpus release (version)";
my $language = $opt_l || die "need to specify a valid language ID";

my $storage_url_base = 'https://object.pouta.csc.fi/OPUS-';

my $file = shift(@ARGV);
my $sent2id = MCDB_File->TIEHASH($database) or die "cannot tie $database ($!)\n";
$sent2id->madvise(MCDB_File::MADV_RANDOM);

my $data_url  = $storage_url_base . $corpus . '/' . $release . '/raw/' . $language . '.zip';
my $data_file = $corpus . '_' . $release . '_raw_' . $language . '.zip';


if (not -e $data_file){
    print STDERR "Downloading $data_url\n";
    system("wget -qq -O $data_file $data_url");
}
    
my $lzip = new Archive::Zip::SimpleUnzip $data_file
    or die "Cannot open zip file: $SimpleUnzipError\n" ;
 

my $parser = new XML::Parser(Handlers => {Start => \&start_element,
					  End => \&end_element,
					  Default => \&char_data});


my ($count, $inSent, $sentStr, $sentID, $sentCount, $parCount, $parID, $document);

my %parBoundaryTags = ('p' => 1,
		       'h1' => 1,
		       'h2' => 1,
		       'h3' => 1,
		       'div' => 1);

while (my $member=$lzip->next){
    my $filename = $member->name();
    if ($filename=~/\.xml$/){

	$inSent    = 0;
        $sentStr   = '';
        $sentID    = '';
        $sentCount = 0;
        $parCount  = 0;
        $parID     = 0;

	my @path = split(/\//,$filename);
	splice(@path,0,2);
	$document  = join('/',@path);

	print STDERR "process $filename ($count sentences done)\n";

	my $fh = $member->open();
	$parser->parse($fh);	
        $count += $sentCount;
    }
}

print STDERR "A total of $count sentences found\n";





sub start_element{
    my ($p, $name, %attr) = @_;

    if ($name eq 's'){
        $inSent = 1;
        $sentCount++;
        $sentStr = '';
	if (exists $attr{'id'}){
            $sentID = $attr{'id'};
	}
	else{
            $sentID = $sentCount;
	}
        if (not $sentCount % 2000){
            print STDERR '.';
	    if (not $sentCount % 100000){
		print STDERR " $sentCount\n";
	    }
	}
    }
    elsif ($parBoundaryTags{lc($name)}){
	$parCount++;
        if (exists $attr{'id'}){
            $parID = $attr{'id'};
	}
        else{
            $parID = $parCount;
	}
    }

}


sub end_element{
    my ($p, $name) = @_;

    if ($name eq 's'){
	$sentStr=~s/^\s*//;
	$sentStr=~s/\s*$//;

	if ($sent2id->EXISTS($sentStr)) {
	    my $id = $sent2id->FETCH($sentStr);
	    print join("\t",($id,$corpus,$document,$parID,$sentID,length($sentStr))),"\n";
	}
	else{
	    print STDERR "cannot find sentence '$sentStr'\n";
	}

        $sentStr = '';
        $inSent = 0;
    }
    elsif ($parBoundaryTags{lc($name)}){
	$parID = 0;
    }

}


sub char_data{
    my ($p, $data) = @_;
    if ($inSent){
        $sentStr .= $data;
    }
}

